File: C#/Euler.cs
/* Euler.cs
*
*  Created on : May 7, 2017
* Author : CleverChuk
* Description:
*   Implementation of Euler's Method for first ODEs
*/
using System;
namespace Euler
{
    public class Euler : IEuler
    {
        public double TOL { get; set; } = 1e-4;

        public double[] solver(fPointer fptr, double t0, double y0, 
            double h, double t, int maxIter)
        {
            double[] yvalues = new double[maxIter];
            yvalues[0] = y0;

            for (int i = 1; i < maxIter; i++)
            {
                yvalues[i] = (yvalues[i - 1] + h * fptr(t0, yvalues[i - 1]));
                t0 += h;

                if(Math.Abs(t0-t)< TOL)
                {
                    break;
                }
            }
            return yvalues;
        }
    }
}

File: C++/Euler.cpp
/* Euler.cpp
*
*  Created on : May 4, 2017
* Author : CleverChuk
*/
#include "Euler.h"
using namespace std;
/*  solver that performs the Euler discretization and approximation of first
	order ODE
*/
vector<double> Euler::solver(double(*fptr)(double t0, double f0), double t0, 
	                 double f0, double h, double t, int maxIter, double tol) {

	vector<double> yvalues(maxIter);
	yvalues[0] = f0 ;
	for (int i = 1; i < maxIter; i++)
	{
		if (abs(t0-t) < tol)
			break;
		
		yvalues[i] = yvalues[i - 1] + h * fptr(t0, yvalues[i - 1]);
		t0 += h;
	}
	
	return yvalues;
}

Euler::Euler()
{
}

Euler::~Euler()
{
}
